lib/spack/spack/caches.py

```diff
@@ -51,24 +55,43 @@ def _fetch_cache():
 
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
         self.skip_unstable_versions = skip_unstable_versions
 
     def store(self, fetcher, relative_dest):
         """Fetch and relocate the fetcher's target into our mirror cache."""
 
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
-        mkdirp(os.path.dirname(dst))
-        fetcher.archive(dst)
+        dst = spack.util.url.join(self.root, relative_dest)
+        local_dst = spack.util.url.local_file_path(dst)
+        if local_dst:
+            mkdirp(os.path.dirname(local_dst))
+            fetcher.archive(local_dst)
+        else:
+            # In the case of non-local destination, fetch to tmp dir, upload to
+            # remote destination, and then delete tmp dir
+            tmpdir = tempfile.mkdtemp(prefix='spack-mirror-cache-store')
+            try:
+                local_dst = os.path.join(tmpdir, relative_dest)
+                mkdirp(os.path.dirname(local_dst))
+                fetcher.archive(local_dst)
+
+                spack.util.web.push_to_url(local_dst, dst, keep_original=False)
+            finally:
+                shutil.rmtree(tmpdir, ignore_errors=True)
```

Comment:
I'm not sure this is necessary, the `fetcher.archive()` call itself seems to do the right thing without having to handle remote cases specially. Local directories get created automatically for local paths and for remote paths the fetcher knows how to push the file that already exists locally to the url. Look at the `archive()` method for the `URLFetchStrategy` and the implementation of `push_to_url()` in `web.py`.

I think the issue before was that this always forced a local path instead of just attempting to store the file in `dst` directly. Did the tests fail without the extra logic to localize the path?

Reply:
I'm pretty sure the tests did fail, though I should double check it, again.  I'm a bit out of my element since I didn't introduce the `MirrorCache` class and it's not very clear to me exactly what its purpose is.

My current understanding is that it is an additional abstraction layer that is supposed to cache downloads from remote mirrors, yet it was clearly not written to handle non-local URLs, so I'm really confused as to what the original intent was.  I'd imagine a caching layer would somehow map non-local mirror URLs to some root directory under the Spack cache, and serve subsequent fetch requests out of that cache, but I don't see that logic anywhere.

I'll take another look and see if I can't make two cents of what this class does, and what needs to change to keep it from bailing on remote URLs.

Reply:
@scheibelp it looks like you were the one who introduced this class, would you mind enlightening us on the intent of it?